Make Measurer generic

kjanosz opened the pull request:
Make Measurer generic.
Refactor code.
Add Pickler based serialization for the measurements.

kjanosz commented on scalameter-core/src/main/scala/org/scalameter/Measurer.scala:
```diff
@@ -279,21 +308,22 @@ object Measurer {
    *  the confidence intervals.
    */
-  trait Noise extends Measurer {
+  trait Noise extends Measurer[Double] {
```
It's a part of Measurer still, but I think that the next step (when splitting `Measurer` to `Runner` and `Gauge` and enabling composition) should be to make it as a function that can be used in the `measurer.map`.
What do you think about it?

axel22 replied:
I guess that would be one way to do it. I don't think it's a high priority, though. It's possible that we should eventually deprecate `Noise` measurer altogether.
And I'd still avoid the `Runner`/`Gauge` split.

kjanosz commented on scalameter-core/src/main/scala/org/scalameter/package.scala:
```diff
@@ -19,6 +19,18 @@ package object scalameter extends MeasureBuilder[Unit, Double](
   MeasureBuilder.average
 ) {
 
+  implicit def inline[T: Ordering](measurer: Measurer[T]): MeasureBuilder[Unit, T] = {
```
I've added it to enable inline benches for all kind of measurers. I suppose that the alternative could be to make `MeasureBuilder.withMeasurer` generic, but that would require to either change default aggregator to median or enforce changing of measurer and aggregator in one method.

axel22 replied:
It seems to me that we should add an overloaded 2-argument `withMeasurer` method that takes:
- a measurer of a different type
- an aggregator/`resultFunction` of that same type

    def withMeasurer[U](m: Measurer[U], f: Seq[Quantity[U]] => Quantity[U]): MeasureBuilder[Unit, U]

This method should live on the `MeasureBuilder` class.

kjanosz replied:
Hm, it looks like signature should look like:

    def withMeasurer[V, M <: Measurer[V]](m: M, a: Seq[V] => V) // this M <: Measurer[V] is needed because Measurer is invariant and can't be made covariant

and there `test("Should correctly measure gc cycles") ` from `InlineBenchmarkTest` should look like this:
    
    withMeasurer(new Measurer.GarbageCollectionCycles, Aggregator.median[Int]) // manually specifying type to `Aggregator.median` is needed because otherwise wrong type is inferred

Or do you have other idea?

kjanosz commented on src/main/scala/org/scalameter/BasePerformanceTest.scala:
```diff
@@ -2,13 +2,17 @@ package org.scalameter
 
 abstract class BasePerformanceTest {
 
+  type Result
```
I've used type alias, since that allows us to specify easily default measurer, aggregator, and so on in the `BasePerformanceTest` descendants.
One caveat: I think that Java users should have different API, but I don't know currently how to do it other way.

axel22 replied:
I think that we should have this as a type parameter on the `BasePerformanceTest` class.
This will break a lot of things, but since it's such a feature-rich release, our clients should be willing to bite the bullet.

kjanosz replied:
OK, but won't it implicate that we won't be able to provide default measurer, reporter and aggregator in the Bench implementation? Or other way, if we provide them and stick to Double user won't be able to use other measurers instead.

axel22 replied:
We have specific `Bench` subclasses that assume a specific type like `Double`. For those, we can specify the default `measurer`, `warmer`, `reporter`, etc.

If there is a non-`Double` measurer, such as the method invocation measurer, then it should have its own `Bench` subclass with default values. And we can only provide it after we have the right reporter, anyway.

kjanosz commented on src/main/scala/org/scalameter/Measurement.scala:
```diff
@@ -1,15 +1,19 @@
 @SerialVersionUID(-2541697615491239986L)
-case class Measurement(value: Double, params: Parameters, data: MeasurementData, units: String) {
-  def complete: Seq[Double] = data.complete
+case class Measurement[T: Pickler](value: T, params: Parameters,
```
`Measurer` is now serialized in json based serializers using `Pickler`s, since in theory T can be arbitrary type so Jackson would have a problem (de)serializing it. 
I guess that doing it purely with Jackson would require specifying e.g. type hints or adding some annotations. So that overall looks like more complex solution than just using picklers.

axel22 replied:
Agreed - using picklers looks like a simpler solution here, consistent with what we were doing before.

kjanosz commented on src/main/scala/org/scalameter/Measurement.scala:
```diff
@@ -1,15 +1,19 @@
   def success: Boolean = data.success
-  def errors: Errors = new Errors(this)
-  def failed = this.copy(data = data.copy(success = false))
+  def errors(implicit num: Numeric[T]): Errors[T] = new Errors(this)
```
I've embedded Numeric as an implicit in the `errors` method and not in the `Measurement` since that should be more robust when dealing with more complex types, like Map[String, Double]
However, this method isn't currently used, so maybe we could remove it?

axel22 replied:
It looks like `errors` is not used anywhere.
Having `errors` in fact pollutes everything with the `Numeric` typeclass.

I would just remove it altogether. Those values in `errors` would only be used by warmers, measurers and reporters that know that the result type is numeric, so they would have the numeric typeclass anyway.

kjanosz commented on src/main/scala/org/scalameter/Persistor.scala:
```diff
@@ -6,15 +6,15 @@ package org.scalameter
-  def load(context: Context): History
-  def save(context: Context, h: History): Unit
+  def load[T: Manifest](context: Context): History[T]
+  def save[T](context: Context, h: History[T]): Unit
```
I haven't made `Persistor` generic since that would be more painful to the users when they would like to change default `Measurer` (need of overriding `def persistor: Persistor[T]` too).
The drawback of it is that only Jackson needs `Manifest` to deserialization, but in the above approach all the reporters that deals with persisting data require `Manifest` to call `persistor.load` even if `persistor` is in fact `SerializationPersistor`.

axel22 replied:
I did not dive into the details, but: sounds like an alternative would be to embed the manifest information about `T` into the `History` class, so that when it's deserialized, Jackson knows how to select the right pickler for its values.

Would that work?

kjanosz replied:
Well, we have already `pickler` embeded in the `Measurement` so I suppose that we could simply read it using `Jackson` as a `History[_]]` and than cast it to the desired type, but I will check that if any problem could occur.

axel22 replied:
Cool - thanks!

axel22 replied:
Btw, along that line - any thoughts about how that plays with the `include` statement?

kjanosz replied:
How does the `include` statement should work? 
I'm asking because I've run the `org.scalameter.examples.TestSuite` and surprisingly only second `include` was executed. It's not the issue with this PR, because I've run it also on the `master` branch and got the same result.

axel22 replied:
It's possible that the `include` statement got broken at some point.

The bench that has the `include` statement should execute all the snippets from the included benchmarks. If there are multiple `include` statements, then all the tests from the sub-benchmarks should be included.

With the current refactoring that introduces bench result types, and the recent mailing list discussion, we would need to think about how we want to execute those nested tests and report their results.
I'm fine if we fix all that later, in a separate PR.

axel22 commented on scalameter-core/src/main/scala/org/scalameter/InvocationCount.scala:
```diff
@@ -3,29 +3,24 @@ package org.scalameter
  *  method invocation counting.
  */
-trait InvocationCount extends Measurer {
+trait InvocationCount extends Measurer[Double] {
```
I expected that the measurer would produce `Map[String, Int]` here?
Then:
- clients can use the `map` operation to extract desired information from this map, such as the total count across all methods
- you could also provide factory methods on the `InvocationCount` companion object to return the total method invocation count

I guess your plan is to do this refactoring in the next PR?

kjanosz replied:
Yes, the plan is to make it in the next PR, although I suppose that it would be handy also if we have the reporter that collects such map result in html table? (I didn't start working on it yet)

axel22 replied:
I would not yet worry about such a reporter - that would be an extra feature. As long as the clients can easily create a `Double` invocation count measurer whose outputs existing reporters can handle, we should be fine.

Then, once we're done with this PR, we can optionally add a new html reporter that shows the breakdown in an html table.

kjanosz replied:
OK, so I will refactor `InvocationCount` in this PR then.

axel22 commented on scalameter-core/src/main/scala/org/scalameter/InvocationCount.scala:
```diff
@@ -36,14 +31,17 @@ trait InvocationCount extends Measurer {
       tear(value)
 
-      invocationsBuilder += MethodInvocationCounter.counts.sum
+//      invocations += Quantity(
```
I guess we want to remove the commented code.

kjanosz replied:
Ups, sorry, I've missed it

axel22 commented on scalameter-core/src/main/scala/org/scalameter/MeasureBuilder.scala:
```diff
@@ -8,21 +8,26 @@ package org.scalameter
   val setup: T => Unit,
   val teardown: T => Unit,
-  val resultFunction: Seq[Double] => U
+  val resultFunction: Seq[U] => U
```
It seems to me that we actually might want to have a `Seq[Quantity[U]]` here. Otherwise, there is no way to take the unit into account when writing the `resultFunction`. It's true that the workaround would be to use a `map` in the measurer to set the unit appropriately, but it would be more flexible to make the `resultFunction` take the full `Quantity`.

axel22 commented on scalameter-core/src/main/scala/org/scalameter/MeasureBuilder.scala:
```diff
@@ -38,12 +43,12 @@ class MeasureBuilder[T, U](
       )
   
-      resultFunction(measurements)
+      resultFunction(measurements.map(_.value))
```
If you refactor `resultFunction` to `Seq[Quantity[U]] => Quantity[U]`, then there would be no need for a `map` here.

kjanosz replied:
Look like the `resultFunction` is simply `Aggregator` now so do you also suggest to change `Aggregator` `extends` from `Seq[T] => T` to `Seq[Quantity[T]] => Quantity[Q]`?

axel22 replied:
I guess so. Do you see any problem with that?

kjanosz replied:
No, I was asking  just for clarification :-)

kjanosz replied:
Ach, I forgot about it - but yes, it looks like some problem can arise. I had a problem earlier with serializing `Quantity` when using `SeparateJvmsExecutor`.

axel22 replied:
That is very weird - I'm not sure why this would happen.
And you are sure that you cleared the temporary folder before you did the change?

kjanosz replied:
Yes, I've cleaned the project and removed `target/benchmarks`. 
I'm wondering if acceptable would be to make it that way:
- unwrap values and units  right after call to the `measure` (same as it's here)
- leave `Aggregator` as function from `Seq[T] => T`
- provide implicit function that lifts `Aggregator` to the `Seq[Quantity[T]] => Quantity[T]`
We wouldn't struggle with this weird error and users would get the possibility to both use aggregators and get value with units in the inline benchmarking.

What do you think about it?

axel22 replied:
What you suggest could be a way. How would you lift `Seq[T] => T` to `Seq[Quantity[T]] => Quantity[T]` without knowing the unit, though?

I still don't get what is causing the error... And you say you deleted both `target` directories in the root, and the `scalameter-core`?

kjanosz replied:
Ach, right but also no need for lift here. We should simply construct `Quantity` from result of aggregation and units from let's say first (or last) from the result sequence (obtained from `measurer.measure`).

Yes, I've cleared both `target` directories.

axel22 replied:
I guess we can do it that way. I don't foresee a need for the aggregator to have access to units, so it could work out.

I'd still be happier if we knew the reason for the error. But serialization stack traces are always fun to deal with..

kjanosz replied:
Hm, I've dropped `Quantity` to tuple right after call to the `measurer.measure` and everything got serialized and deserialized properly.
So I think that this would be better workaround, but I still don't know the exact cause and it's weird to me that tuples work but custom `Quantity` class not.
Btw, I've tried to drop usage of `case class` here and mixing `Serializable` manually but that didn't work too.

axel22 replied:
Not sure what's the reason... Well, let's do it like this then. We'll debug serialization some other time :)

axel22 commented on scalameter-core/src/main/scala/org/scalameter/Measurer.scala:
```diff
@@ -52,12 +76,12 @@ object Measurer {
-  trait Timer extends Measurer {
-    def units = "ms"
+  trait Timer extends Measurer[Double] {
+//    def units: UnitOfMeasure[Double] = Time.Milliseconds
```
I guess that we want to remove this comment.

axel22 commented on scalameter-core/src/main/scala/org/scalameter/Measurer.scala:
```diff
@@ -72,22 +96,19 @@ object Measurer {
+      tear: T => Any, regen: () => T, snippet: T => Any): Seq[Quantity[Double]] = {
       var iteration = 0
-      var times = List[Double]()
+      val times = mutable.ListBuffer.empty[Quantity[Double]]
```
I hope that this does not change the behavior of any measurer.
`OutlierElimination` does sorting, so I guess we should be fine.

kjanosz replied:
I've changed it to be more consistent, but don't know if it could break something:
Earlier:
- time-based measurers were prepending result
- memory and gc count were prepending and then doing `reverse` at the end

axel22 replied:
I think it should be fine.

axel22 commented on scalameter-core/src/main/scala/org/scalameter/Measurer.scala:
```diff
@@ -251,19 +279,20 @@ object Measurer {
+      while (outlierExists(results.map(_.value.toDouble())) && retleft > 0) {
+        val prefixlen = suffixLength(results.reverse.map(_.value.toDouble()))
+        val suffixlen = suffixLength(results.map(_.value.toDouble()))
+        val formatted = results.map(t => f"${t.value.toDouble()}%.3f")
```
Any reason not to use `t.toString` instead of converting it to `Double` first?

kjanosz replied:
Well, if this generic param is `Double` and those `Double`  have a lot of digit after decimal point then the printing won't be so readable.
On the other hand, yes we unnecessarily also convert `Int`s and so on to `Double`.
So I don't know which choice is better.
Alternatively, I guess that some pattern matching here could be an option.

axel22 replied:
But will it be readable if we convert `Double` to `Double`? Won't the output in case of a lot of decimals look the same?

But, I guess this is not a big issue - I'm fine with either solution.

kjanosz replied:
`Numeric[Double].toDouble(d)` returns value itself and also won't f"num$.3f" formatter guarantee us that any `Double` will be formatted to max three digits after decimal point? Also note, that such formatter requires `Double` values.

axel22 replied:
I missed that - I'm fine with the current solution then.

axel22 commented on src/main/scala/org/scalameter/Executor.scala:
```diff
@@ -1,26 +1,28 @@
     for (setup <- setuptree) yield {
-      val exec = Option(setup.customExecutor).getOrElse(this)
-      val cd = exec.runSetup(setup)
+      val cd = runSetup(setup)
```
Interesting. A quick search through the code says we're not using the `customExecutor` field any more. This means that the change in this line should not break anything. I see that you also removed it below - nice catch.

axel22 commented on src/main/scala/org/scalameter/Executor.scala:
```diff
@@ -30,21 +32,22 @@ object Executor {
-  trait Factory[E <: Executor] {
-    def apply(warmer: Warmer, aggregator: Aggregator, m: Measurer): E
+  trait Factory[E[_] <: Executor[_]] {
+    def apply[T: Numeric: Pickler](warmer: Warmer,
```
Since `aggregator` is coming here externally, it seems to me that we don't need a `Numeric` typeclass here, or the `Pickler` - they should be provided at the instantiation site when the `aggregator` and `Measurer` are created. Or am I missing something?

kjanosz replied:
About the `Numeric` - agreed, but `Pickler` is unfortunately needed. Both `LocalExecutor` and `SeparateJvmsExecutor` are creating `Measurement` so they require `Pickler` to be available in the scope.

kjanosz replied:
And also, `Numeric` is needed is `SeparateJvmsExecutor` to convert results to `Double` and print them (in the `nice` function) - please see [my previous comment](#r37221398).

axel22 replied:
I agree about the `Pickler` then, but I still think the case for `Numeric` is not strong enough, particularly because that type parameter will not always refer to a numeric type.

It would be better to introduce your own `PrettyPrinter` typeclass, which is derived directly from numeric for numeric types:

    implicit def numericPrettyPrinter[T: Numeric] = ...
    implicit def defaultPrettyPrinter[T] = ...

axel22 commented on src/main/scala/org/scalameter/Reporter.scala:
```diff
@@ -2,26 +2,25 @@ package org.scalameter
-  object None extends Reporter {
-    def report(result: CurveData, persistor: Persistor) {}
-    def report(results: Tree[CurveData], persistor: Persistor) = true
+  object None extends Reporter[Nothing] {
```
Since `Reporter` is not covariant, we should instead have a method that creates a reporter of the desired type:

    def None[T] = new Reporter[T] { ... }

axel22 commented on src/main/scala/org/scalameter/Setup.scala:
```diff
@@ -8,8 +8,7 @@ case class Setup[T](
   teardown: Option[T => Any],
   customwarmup: Option[() => Any],
-  snippet: T => Any,
-  @transient customExecutor: Executor
```
Nice catch!

axel22 commented on src/main/scala/org/scalameter/execution/LocalExecutor.scala:
```diff
@@ -26,18 +27,17 @@ import utils.Tree
  *  the young generation.
  */
-class LocalExecutor(val warmer: Warmer, val aggregator: Aggregator, val measurer: Measurer) extends Executor {
+class LocalExecutor[V: Numeric: Pickler](val warmer: Warmer, val aggregator: Aggregator[V], val measurer: Measurer[V]) extends Executor[V] {
```
Seems to me that we could get rid off the typeclasses here?

kjanosz commented:
@axel22: Hm, I should have run all tests again locally before rebasing and pushing, but well it looks like there are other serialization problems too. I guess that it's connected with fact that now `Bench` is parametrized.
I will dig into that and try to fix it.

kjanosz commented:
@axel22: Now it should be fixed.
However, there are couple caveats:

- I had to add similar workaround for the `InvocationCount` - instead of returning `Map[MethodSignature, Long]` I currently return `Map[String, Long]`. The problem with it that it looks like we can't use any class from `ScalaMeter`. I don't know what is the exact cause of it, but I suspect some classpath problems.
- It looks like `ScalaMeterFramework` is broken now and users can't run their tests in sbt. I will look into cause of it.

axel22 commented:
About the first bullet, I'm not sure I understand what the reason is - you're saying that if `InvocationCount` is parametrized with `Map[MethodSignature, Long]`, running the test blows up with a `ClassCastException`, but works with `Map[String, Long]`? Weird.

The solution you did for the `ScalaMeterFramework` should work, but I still don't get why checking if class conforms to `BasePerformanceTest` does not work.

kjanosz commented:
Yes  - both (`Quantity` and `MethodSignature`) classes are from the `ScalaMeter`. Also, I've done some tests playing with `MeasurerTest.measureWith` method and it looks like returning e.g. `Quantity[Double]` alone is OK, but returning `Seq[Quantity[Double]]` is not (same apply for the `MethodSignature` and I suppose that for every `ScalaMeter` class).

I will take a look at it again tomorrow with some fresh mind.

kjanosz commented:
@axel22:
About the `ClassNotFoundException` - I completely don't know the cause of it. `ObjectInputStream.resolveClass` calls `Class.forName` with `VM.latestUserDefinedLoader`argument, which contains `ScalaMeter` classes (both `scalameter-core` and `scalameter`), so everything should be OK (well, but it isn't).

About the sbt integration:
I suppose that it doesn't work because `xsbt.api.Discovery.simpleName` returns `None` for the `xsbti.api.Polymorphic` (I've made an assumption that generic base class is mapped  to the `xsbti.api.Polymorphic`.)

axel22 commented:
Ok - about `ClassCastException`s - even if we don't understand why this is happening, we should make sure that all ScalaMeter features work before moving forward:
- we should ensure that benchmarks work for all kinds of measurers and executors
- we should ensure that inline benchmarks work for different measurer types

About SBT discovery - if you can solve the issue by introducing another base trait, that is fine.

About introducing defaults for `Bench` and `JBench` - makes sense, but we should do that in a separate PR - this one is quite large already.

kjanosz commented:
@axel22:
I've done some reorganization in measurer tests - I've put all example benches in the `org/scalameter/examples/measurersTests.scala` file. I've moved `BoxingCountTest` and `MethodInvocationCountTest` to one `org/scalameter/invocationCountMeasurersTests.scala` file and placed tests for rest of benchmarks in the `org/scalameter/measurersTests.scala`.

Inline benchmarks are already tested in `InlineBenchmarkTest` which covers also measurer of other type - gc count, which is `Measurer[Int]`.

About the sbt discovery:
Yes, I've made such change and introduced `AbstractPerformanceTest` and tested in on separate project and everything seems to work.

axel22 commented:
Ok, nice work! I'm merging this in, then.
